compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -2514,17 +2570,132 @@ private boolean doesRecordContainKeys(BType varType,
+        }
+        return this.types.mergeTypes(restVarSymbolMapType, restConstraintType);
+    }
+
+    BRecordType createRecordTypeForRestField(Location pos, SymbolEnv env, BRecordType recordType,
+                                       List<String> variableList,
+                                       BType restConstraint) {
+        BRecordTypeSymbol recordSymbol = createAnonRecordSymbol(env, pos);
+        BRecordType recordVarType = new BRecordType(recordSymbol);
+        LinkedHashMap<String, BField> unMappedFields = new LinkedHashMap<>() {{
+            putAll(recordType.fields);
+            if(recordType.restFieldType.tag == TypeTags.RECORD) {
+                putAll(((BRecordType) recordType.restFieldType).fields);
+            }
+        }};
+
+        setRestRecordFields(pos, env, unMappedFields, variableList, restConstraint, recordVarType);
+
+        BLangRecordTypeNode recordTypeNode = TypeDefBuilderHelper.createRecordTypeNode(recordVarType,
+                env.enclPkg.packageID, symTable, pos);
+        recordTypeNode.initFunction =
+                TypeDefBuilderHelper.createInitFunctionForRecordType(recordTypeNode, env, names, symTable);
+        TypeDefBuilderHelper.addTypeDefinition(recordVarType, recordSymbol, recordTypeNode, env);
+
+        return recordVarType;
+    }
+
+    void setRestRecordFields(Location pos, SymbolEnv env,
+                             LinkedHashMap<String, BField> unMappedFields,
+                             List<String> variableList, BType restConstraint,
+                             BRecordType targetRestRecType) {
+        LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+        LinkedHashMap<String, BField> markAsOptional = new LinkedHashMap<>();
+
+        if (!targetRestRecType.fields.isEmpty()) {
+            fields.putAll(targetRestRecType.fields);
+            List<String> intersectionFields = getIntersectionFields(Arrays.asList(targetRestRecType.fields,
+                    unMappedFields));
+            markAsOptional.putAll(unMappedFields);
+            markAsOptional.putAll(targetRestRecType.fields);
+            intersectionFields.forEach(markAsOptional::remove);
+        }
+
+        //adds a never-typed optional field for the mapped bindings
+        for (String fieldName : variableList) {
+            unMappedFields.remove(fieldName);
+            BField newField = new BField(names.fromString(fieldName), pos,
+                    new BVarSymbol(Flags.OPTIONAL, names.fromString(fieldName), env.enclPkg.symbol.pkgID,
+                            symTable.neverType, targetRestRecType.tsymbol, pos, VIRTUAL));
+            fields.put(fieldName, newField);
+        }
+
+        for (BField field : unMappedFields.values()) {
+            if (fields.containsKey(field.name.value)) {
+                BField targetField = fields.get(field.getName().value);
+                targetField.type = types.mergeTypes(targetField.type, field.type);
+            } else {
```

Comment:
There seems to be a codecov comment? Is this not covered by tests?

Reply:
Not sure why this comment was added. The specific lines are covered from `RecordVariableDefinitionTest` junit test